File: src/main/java/com/ceraia/arena/listeners/ArenaFightListener.java
package com.ceraia.arena.listeners;

import com.ceraia.arena.Double;
import com.ceraia.arena.managers.InviteManager;
import com.ceraia.arena.types.Arena;
import net.kyori.adventure.text.minimessage.MiniMessage;
import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.Listener;
import org.bukkit.event.entity.EntityDamageByEntityEvent;
import org.bukkit.event.entity.EntityDamageEvent;
import org.bukkit.event.entity.PlayerDeathEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.bukkit.scheduler.BukkitRunnable;

import java.util.Objects;
import java.util.UUID;

public class ArenaFightListener implements Listener {

    private final Double plugin;

    public ArenaFightListener(Double plugin) {
        this.plugin = plugin;
        Bukkit.getPluginManager().registerEvents(this, plugin);
    }

    private boolean isInArena(Player player) {
        return plugin.getArenaManager().getArena(player) != null;
    }

    @EventHandler
    public void onHit(EntityDamageByEntityEvent e) {
        if (e.getDamager() == null || e.getEntity() == null) {
            return;
        }

        if (!(e.getDamager() instanceof Player damager) || !(e.getEntity() instanceof Player player)) {
            return;
        }

        if (!isInArena(damager) && !isInArena(player)) {
            return;
        }

        if (((isInArena(damager) && !isInArena(player)) || (!isInArena(damager) && isInArena(player))) || !Objects.equals(plugin.getArenaManager().getArena(damager).getName(), plugin.getArenaManager().getArena(player).getName())) {
            e.setCancelled(true);
            return;
        }

        Arena arena = plugin.getArenaManager().getArena(damager);

        if (arena == null) {
            return;
        }

        if (arena.getState() != Arena.ArenaState.RUNNING) {
            e.setCancelled(true);
            return;
        }

        if (arena.getTeam1().contains(damager) && arena.getTeam1().contains(player)) {
            e.setCancelled(true);
            return;
        }

        if (arena.getTeam2().contains(damager) && arena.getTeam2().contains(player)) {
            e.setCancelled(true);
        }
    }

    @EventHandler
    public void onDamage(EntityDamageEvent e) {
        if (!(e.getEntity() instanceof Player player)) {
            return;
        }

        if (!isInArena(player)) {
            return;
        }

        // Get player that hurt the player
        Player killer = null;
        if (e instanceof EntityDamageByEntityEvent event) {
            // If the damage is caused by a player
            if (event.getDamager() instanceof Player) {
                plugin.getLogger().info("Player");
                killer = (Player) event.getDamager();
            }
            // If the damage is caused by a projectile
            else if (event.getDamager() instanceof org.bukkit.entity.Projectile projectile) {
                plugin.getLogger().info("Arrow");
                if (projectile.getShooter() instanceof Player) {
                    killer = (Player) projectile.getShooter();
                }
            }

            // If the damage is caused by a tnt
            else if (event.getDamager().getType() == org.bukkit.entity.EntityType.PRIMED_TNT) {
                plugin.getLogger().info("TNT");
                if (event.getDamager().customName() != null) {
                    killer = Bukkit.getPlayer(Objects.requireNonNull(event.getDamager().customName()).toString());
                }
            }
        }

        Arena arena = plugin.getArenaManager().getArena(player);

        double healthAfter = player.getHealth() - e.getFinalDamage();
        if (healthAfter <= 0) {

            // Check if during the fight totems are allowed

            InviteManager.Invite invite = plugin.getInviteManager().invites.get(player);

            if (invite == null) invite = plugin.getInviteManager().selectingInvites.get(killer);

            if (invite != null) {
                if (arena.totems) {
                    if (player.getInventory().getItemInMainHand().getType() == Material.TOTEM_OF_UNDYING || player.getInventory().getItemInOffHand().getType() == Material.TOTEM_OF_UNDYING) {
                        return;
                    }
                }
            }


            e.setCancelled(true);
            player.setHealth(player.getHealthScale());

            arena.end(player, false);
        }
    }

    @EventHandler
    public void onDeath(PlayerDeathEvent e) {
        if (!isInArena(e.getEntity())) {
            return;
        }
        Arena arena = plugin.getArenaManager().getArena(e.getEntity());

        Location loc = e.getEntity().getLocation();

        Player killer = e.getEntity().getKiller();

        if (killer == null) {
            killer = Bukkit.getPlayer(Objects.requireNonNull(Objects.requireNonNull(e.getEntity().getLastDamageCause()).getEntity().customName()).toString());
        }
        if (killer != null) {
            matchEnd(e.getEntity(), killer);
        }



        e.getEntity().spigot().respawn();
        new BukkitRunnable() {
            @Override
            public void run() {
                e.getEntity().teleport(loc);
                arena.end(e.getEntity(), false);
            }
        }.runTaskLater(plugin, 5L);
    }

    @EventHandler
    public void onQuit(PlayerQuitEvent e) {
        if (!isInArena(e.getPlayer())) {
            return;
        }

        Arena arena = plugin.getArenaManager().getArena(e.getPlayer());

        // Check in which team the player is
        if (arena.getTeam1().contains(e.getPlayer())) {
            matchEnd(e.getPlayer(), arena.getTeam2().get(0));
        } else {
            matchEnd(e.getPlayer(), arena.getTeam1().get(0));
        }

        arena.end(e.getPlayer(), true);
    }

    private void matchEnd(Player loser, Player winner) {
        UUID winnerUUID = winner.getUniqueId();
        UUID loserUUID = loser.getUniqueId();

        // Get the win chance
        int winChance = plugin.getPlayerManager().CalculateWinChance(winnerUUID, loserUUID);

        // Announce the winner and the win chance in chat
        Bukkit.broadcast(MiniMessage.miniMessage().deserialize(plugin.getConfig().getString("messages.fight.end_global").replace("%winner%", winner.getName()).replace("%loser%", loser.getName()).replace("%elo%", String.valueOf(plugin.getPlayerManager().getDoublePlayer(loserUUID).getElo())).replace("%winchance%", String.valueOf(winChance)).replace("%arena%", plugin.getArenaManager().getArena(loser).getName()))

        );

        // Handle ELO calculations
        plugin.getPlayerManager().PlayerKill(winnerUUID, loserUUID);
    }
}

File: src/main/java/com/ceraia/modules/arenas/listeners/ArenaFightListener.kt
package com.ceraia.modules.arenas.listeners

import com.ceraia.Ceraia
import com.ceraia.modules.arenas.types.Arena
import net.kyori.adventure.text.minimessage.MiniMessage
import org.bukkit.Bukkit
import org.bukkit.Material
import org.bukkit.entity.*
import org.bukkit.event.EventHandler
import org.bukkit.event.Listener
import org.bukkit.event.entity.EntityDamageByEntityEvent
import org.bukkit.event.entity.EntityDamageEvent
import org.bukkit.event.entity.PlayerDeathEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.bukkit.scheduler.BukkitRunnable
import java.util.*

class ArenaFightListener(private val plugin: Ceraia) : Listener {
    init {
        Bukkit.getPluginManager().registerEvents(this, plugin)
    }

    private fun isInArena(player: Player): Boolean {
        return plugin.arenaModule.arenaManager?.getArena(player) != null
    }

    @EventHandler
    fun onHit(e: EntityDamageByEntityEvent) {
        val damager = e.damager as? Player ?: return
        val player = e.entity as? Player ?: return

        if (!isInArena(damager) && !isInArena(player)) {
            return
        }

        if (((isInArena(damager) && !isInArena(player)) || (!isInArena(damager) && isInArena(player))) || plugin.arenaModule.arenaManager?.getArena(
                damager
            )?.name != plugin.arenaModule.arenaManager?.getArena(player)?.name
        ) {
            e.isCancelled = true
            return
        }

        val arena: Arena = plugin.arenaModule.arenaManager?.getArena(damager) ?: return

        if (arena.state != Arena.ArenaState.RUNNING) {
            e.isCancelled = true
            return
        }

        if (arena.team1.contains(damager) && arena.team1.contains(player)) {
            e.isCancelled = true
            return
        }

        if (arena.team2.contains(damager) && arena.team2.contains(player)) {
            e.isCancelled = true
        }
    }

    @EventHandler
    fun onDamage(e: EntityDamageEvent) {
        val player = e.entity as? Player ?: return

        if (!isInArena(player)) {
            return
        }

        // Get player that hurt the player
        var killer: Player? = null
        if (e is EntityDamageByEntityEvent) {
            // If the damage is caused by a player
            if (e.damager is Player) {
                plugin.logger.info("Player")
                killer = e.damager as Player
            } else if (e.damager is Projectile) {
                plugin.logger.info("Arrow")
                val projectile = e.damager as Projectile
                if (projectile.shooter is Player) {
                    killer = projectile.shooter as Player
                }
            } else if (e.damager.type == EntityType.TNT) {
                plugin.logger.info("TNT")
                if (e.damager.customName() != null) {
                    killer = Bukkit.getPlayer(Objects.requireNonNull(e.damager.customName()).toString())
                }
            }
        }

        val arena: Arena = plugin.arenaModule.arenaManager?.getArena(player) ?: return

        val healthAfter: Double = player.health - e.finalDamage
        if (healthAfter <= 0) {
            // Check if during the fight totems are allowed

            var invite = plugin.arenaModule.inviteManager!!.invites[player]

            if (invite == null) invite = plugin.arenaModule.inviteManager!!.selectingInvites[killer]

            if (invite != null) {
                if (arena.totems) {
                    if (player.inventory.itemInMainHand.type == Material.TOTEM_OF_UNDYING || player.inventory.itemInOffHand.type == Material.TOTEM_OF_UNDYING) {
                        return
                    }
                }
            }

            e.isCancelled = true
            player.health = player.healthScale

            arena.end(player, false)
        }
    }

    @EventHandler
    fun onDeath(e: PlayerDeathEvent) {
        if (!isInArena(e.entity)) {
            return
        }
        val arena = plugin.arenaModule.arenaManager?.getArena(e.entity) ?: return

        val loc = e.entity.location

        var killer = e.entity.killer

        if (killer == null) {
            killer = Bukkit.getPlayer(
                Objects.requireNonNull(Objects.requireNonNull(e.entity.lastDamageCause)?.entity?.customName()).toString()
            )
        }
        if (killer != null) {
            matchEnd(e.entity, killer)
        }

        e.entity.spigot().respawn()
        object : BukkitRunnable() {
            override fun run() {
                e.entity.teleport(loc)
                arena.end(e.entity, false)
            }
        }.runTaskLater(plugin, 5L)
    }

    @EventHandler
    fun onQuit(e: PlayerQuitEvent) {
        if (!isInArena(e.player)) {
            return
        }

        val arena = plugin.arenaModule.arenaManager?.getArena(e.player) ?: return

        // Check in which team the player is
        if (arena.team1.contains(e.player)) {
            matchEnd(e.player, arena.team2[0])
        } else {
            matchEnd(e.player, arena.team1[0])
        }

        arena.end(e.player, true)
    }

    private fun matchEnd(loser: Player, winner: Player) {
        val winnerUUID = winner.uniqueId
        val loserUUID = loser.uniqueId

        // Get the win chance
        val winChance = plugin.arenaModule.calculateWinChance(winnerUUID, loserUUID).toInt()

        // Announce the winner and the win chance in chat
        Bukkit.broadcast(
            MiniMessage.miniMessage().deserialize(
                plugin.config.getString("messages.fight.end_global")!!
                    .replace("%winner%", winner.name).replace("%loser%", loser.name)
                    .replace("%elo%", plugin.playerManager.getCeraiaPlayer(loserUUID).elo.toString())
                    .replace("%winchance%", winChance.toString())
                    .replace("%arena%", plugin.arenaModule.arenaManager?.getArena(loser)?.name ?: "")
            )
        )

        // Handle ELO calculations
        // #TODO: Re-implement ELO calculations
        //plugin.getPlayerManager().PlayerKill(winnerUUID, loserUUID);
    }
}